text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+
+- We identified a set of "mainstream" platforms, consisting of 32- and 64-bit
+  machines running Windows, Linux, or macOS. "Portability by default" thus more
+  specifically means portability *to mainstream platforms*.
+
+- We present an ergonomic, primary API surface which is portable across these
+  mainstream platforms (see `std::{fs, net, env, process, sync}` etc.).
+
+- We *also* provide separate access to low-level or OS-specific services via the
+  `std::os` module. APIs in this module are largely traits that extend the
+  cross-platform APIs, and in particular can expose their OS-level
+  representation. The fact that these APIs require explicitly importing from
+  `std::os` provided a small "speed bump" for venturing out of guaranteed
+  mainstream platform portability.
+
+- Finally, for working in low-level and embedded contexts, we stabilized
+  `libcore`, a subset of `libstd` that excludes all OS services and allocation,
+  but *still* makes some hardware assumptions (e.g. about atomics and floating
+  point support).
+
+## Problems with the status quo
+
+The above strategy has served us fairly well in the first year since Rust 1.0,
+but it's increasingly holding us back from enhancements we'd like to make, and
+even for the needs it covers, it's suboptimal in a few ways.
+
+**Problems with `std::os`**:
+
+* The `std::os` module has submodules that correspond to a hierarchy of OS
+  types. For example, there is a `unix` submodule that applies to several
+  operating systems, but there's also a `linux` submodule with Linux-specific
+  extensions. There are a couple of problems with such an organization. Most
+  importantly, it's not at all clear how to use the module hierarchy to organize
+  features like [fixed-size atomic types][more-atomics], where the types
+  available vary in a fine-grained way based on the CPU family; [SIMD] is even
+  worse. But even just for operating systems, organizing into a hierarchy
+  becomes difficult as we gain more and more APIs, some of which are only
+  available on particular *versions* of a given operating system.
+
+* The "speed bump" for using `std::os` is minimal and easy to miss; it's just an
+  import that looks the same as any other. Moreover, it doesn't provide any help
+  with the ecosystem beyond `std`. There's no simple way to tell whether a crate
+  you're relying on is portable to the same degree as `std` is, and the `os`
+  submodule pattern has not really caught on in the wider ecosystem.
+
+* Platform-specific APIs don't live in their "natural location". The majority of
+  `std::os` works through extension traits to enhance the functionality of
+  standard primitives. For example `std::os::unix::io::AsRawFd` is a trait with
+  the `as_raw_fd` method (to extract a file descriptor). If you were to ignore
+  Windows, however, one might expect this API instead to live as a method
+  directly on types like `File`, `TcpStream`, etc. Forcing code to live in
```

Comment:
I don't like the running example of `as_raw_fd` needing to be an inherent method. If you're on unix, it makes perfect sense for all file descriptor wrapper types to implement a common trait. For example, if you're writing a crate that implements select/poll, you might want to be generic over `AsRawFd`.